### ISSUE TITLE
BUG for ONNX inference

### ISSUE THREAD

**Issue by Fred-cell** [wenet-e2e/wenet#1113]
when i inference with u2++_conformer, execute just 50 wav files, a bug will be thrown as below:
I0515 00:10:40.909694 146005 decoder_main.cc:67] num frames 1118
I0515 00:10:41.026697 146005 decoder_main.cc:86] Partial result: 在机关
I0515 00:10:41.056061 146005 decoder_main.cc:86] Partial result: 在机关服务
I0515 00:10:41.085124 146005 decoder_main.cc:86] Partial result: 在机关围剿
I0515 00:10:41.110785 146005 decoder_main.cc:86] Partial result: 在机关围剿和
I0515 00:10:41.136417 146005 decoder_main.cc:86] Partial result: 在机关围剿和
I0515 00:10:41.176227 146005 decoder_main.cc:86] Partial result: 在机关围剿和工程
I0515 00:10:41.217715 146005 decoder_main.cc:86] Partial result: 在机关围剿和工程多处的
I0515 00:10:41.251241 146005 decoder_main.cc:86] Partial result: 在机关围剿和工程多处的战斗中
I0515 00:10:41.282459 146005 decoder_main.cc:86] Partial result: 在机关围剿和工程多处的战斗中太勇敢
I0515 00:10:41.311969 146005 decoder_main.cc:86] Partial result: 在机关围剿和工程多处的战斗中太勇敢坚定
I0515 00:10:41.341024 146005 decoder_main.cc:86] Partial result: 在机关围剿和工程多处的战斗中太勇敢坚定是
I0515 00:10:41.398414 146005 decoder_main.cc:86] Partial result: 在机关围剿和工程多处的战斗中太勇敢坚定是一军的
I0515 00:10:41.429834 146005 decoder_main.cc:86] Partial result: 在机关围剿和工程多处的战斗中太勇敢坚定是一军的
I0515 00:10:41.462321 146005 decoder_main.cc:86] Partial result: 在机关围剿和工程多处的战斗中太勇敢坚定是一军的主要将领
Segmentation fault (core dumped)
this file shoud be processed completely, i will go deep to locate the bug info.

onnxruntime: 1.10.0 and 1.11.1

**Comment by Fred-cell** [comment]
BUG is from inference of decoder part:
std::vector<Ort::Value> rescore_outputs = rescore_session_->Run(
      Ort::RunOptions{nullptr}, rescore_in_names_.data(), rescore_inputs.data(),
      rescore_inputs.size(), rescore_out_names_.data(),
      rescore_out_names_.size());

**Comment by robin1001** [comment]
@lzhin @xingchensong please follow the issue.

**Comment by xingchensong** [comment]
1. please show us `chunk_size` and `num_left_chunks` when u export onnx
2. please show us `chunk_size` and `num_left_chunks` when u call decoder_main
3. please show us ur log info when initializing the decoder

**Comment by Fred-cell** [comment]
export script is as below:
python wenet/bin/export_onnx_cpu.py  --config exp/u2++_conformer/train.yaml --checkpoint exp/u2++_conformer/final.pt --chunk_size -1 --beam 15 --num_decoding_left_chunks -1 --output_dir ./onnx/
exec decoder_main as below:
./build/decoder_main --chunk_size 16  --wav_scp ./wav.scp   --onnx_dir ./onnx --fst_path ./data/lang_test/TLG.fst  --dict_path ./words.txt --beam 15 --rescoring_weight 1 --ctc_weight 0.5 -reverse_weight 0.5

**Comment by xingchensong** [comment]
Please note that when u export onnx model , -1/-1 means non-streaming model and `chunk_size` used in `decoder_main` should be equal to your exported model (which should be -1). If u want to use 16/-1 in `decoder_main`, please export onnx in the same config.

![image](https://user-images.githubusercontent.com/13466943/168476495-093c9462-13b5-4bcd-a021-cc49538cfc4e.png)

**Comment by Fred-cell** [comment]
I try it according to your opinion, and there is still error as before:
python wenet/bin/export_onnx_cpu.py  --config exp/u2++_conformer/train.yaml --checkpoint exp/u2++_conformer/final.pt --chunk_size 16 --beam 15 --num_decoding_left_chunks -1 --output_dir ./onnx/

**Comment by xingchensong** [comment]
Please add `reverse_weight` > 0 when export onnx. The default value of `reverse_weight` in export_onx_cpu is 0.0, which does not match the value used in `decoder_main`

**Comment by xingchensong** [comment]
If u export a u2++ model with `reverse_weight = 0.0`, ONNX will automatically remove the right-to-left decoder, this means that u can never get `r_score` in `decoder_main`, that's why u got a core dump.

**Comment by Fred-cell** [comment]
when i set --reverse_weight 0.5, inference the fist wave file caused error:
python wenet/bin/export_onnx_cpu.py  --config exp/u2++_conformer/train.yaml --checkpoint exp/u2++_conformer/final.pt --chunk_size 16 --beam 15 --num_decoding_left_chunks -1 --reverse_weight 0.5 --output_dir ./onnx/
![image](https://user-images.githubusercontent.com/99886928/168477273-e132b44f-a341-4289-878d-45666b08bbf3.png)

**Comment by xingchensong** [comment]
Pull latest code and try again, see PR #1117

**Comment by Fred-cell** [comment]
I pull the lastest code, and there is another error:
I0515 22:31:05.134997  3119 decoder_main.cc:67] num frames 905
I0515 22:31:05.309218  3119 decoder_main.cc:88] Partial result: 但是他丝毫
I0515 22:31:05.346619  3119 decoder_main.cc:88] Partial result: 但是他丝毫没有
I0515 22:31:05.374295  3119 decoder_main.cc:88] Partial result: 但是他丝毫没有城市
I0515 22:31:05.407676  3119 decoder_main.cc:88] Partial result: 但是他丝毫没有城市娇小姐
I0515 22:31:05.440114  3119 decoder_main.cc:88] Partial result: 但是他丝毫没有城市较小的样子
I0515 22:31:05.479857  3119 decoder_main.cc:88] Partial result: 但是他丝毫没有城市娇小姐样子反而
I0515 22:31:05.511783  3119 decoder_main.cc:88] Partial result: 但是他丝毫没有城市娇小姐样子反而非常的
I0515 22:31:05.545146  3119 decoder_main.cc:88] Partial result: 但是他丝毫没有城市娇小姐样子反而非常的朴实
I0515 22:31:05.580701  3119 decoder_main.cc:88] Partial result: 但是他丝毫没有城市娇小姐样子反而非常的朴实
I0515 22:31:05.601506  3119 decoder_main.cc:81] Rescoring begin
2022-05-15 21:31:05.608420024 [E:onnxruntime:, sequential_executor.cc:368 Execute] Non-zero status code returned while running Split node. Name:'Split_55' Status Message: Cannot split using values in 'split' attribute. Axis=0 Input shape={10} NumOutputs=15 Num entries in 'split' (must equal number of outputs) was 15 Sum of sizes in 'split' (must equal size of selected axis) was 15
terminate called after throwing an instance of 'Ort::Exception'
  what():  Non-zero status code returned while running Split node. Name:'Split_55' Status Message: Cannot split using values in 'split' attribute. Axis=0 Input shape={10} NumOutputs=15 Num entries in 'split' (must equal number of outputs) was 15 Sum of sizes in 'split' (must equal size of selected axis) was 15
Aborted (core dumped)

**Comment by xingchensong** [comment]
Will `beam=10` work?

**Comment by Fred-cell** [comment]
No, I set beam=10, some files can work but this as below:
I0515 22:46:56.649355 50924 decoder_main.cc:67] num frames 891
I0515 22:46:56.827426 50924 decoder_main.cc:88] Partial result: 他坚信
I0515 22:46:56.852329 50924 decoder_main.cc:88] Partial result: 他坚信只要
I0515 22:46:56.878832 50924 decoder_main.cc:88] Partial result: 他坚信只要踏实肯
I0515 22:46:56.921988 50924 decoder_main.cc:88] Partial result: 他坚信只要踏实肯干
I0515 22:46:56.951897 50924 decoder_main.cc:88] Partial result: 他坚信只要踏实肯干一定能够
I0515 22:46:56.979951 50924 decoder_main.cc:88] Partial result: 他坚信只要踏实肯干一定能够上好
I0515 22:46:57.022194 50924 decoder_main.cc:88] Partial result: 他坚信只要踏实肯干一定能够上好日子
I0515 22:46:57.051928 50924 decoder_main.cc:81] Rescoring begin
2022-05-15 21:46:57.053215410 [E:onnxruntime:, sequential_executor.cc:368 Execute] Non-zero status code returned while running Split node. Name:'Split_45' Status Message: Cannot split using values in 'split' attribute. Axis=0 Input shape={6} NumOutputs=10 Num entries in 'split' (must equal number of outputs) was 10 Sum of sizes in 'split' (must equal size of selected axis) was 10
terminate called after throwing an instance of 'Ort::Exception'
  what():  Non-zero status code returned while running Split node. Name:'Split_45' Status Message: Cannot split using values in 'split' attribute. Axis=0 Input shape={6} NumOutputs=10 Num entries in 'split' (must equal number of outputs) was 10 Sum of sizes in 'split' (must equal size of selected axis) was 10

**Comment by xingchensong** [comment]
It's really strange, `beam` is a dynamic axis when exporting model, but in runtime, onnx seems to treat it as a static aixs.

**Comment by Fred-cell** [comment]
Yes, we can submit a issue to onnxruntime, this bug is accidental.

**Comment by Fred-cell** [comment]
when onnxruntime works, i will try inference based OpenVINO. if the performance is good, i will commit a merge request.

**Comment by xingchensong** [comment]
I know why.  https://github.com/wenet-e2e/wenet/blob/main/wenet/transformer/asr_model.py#L669-L684 . ONNX treats `for-loop`  statically,  it will always split out 15 tensors if we use `beam=15` when exporting the model.

Solution-1: Force padding hyps in https://github.com/wenet-e2e/wenet/blob/main/runtime/core/decoder/onnx_asr_model.cc#L315-L330 if `num_hyps < predefined beamsize`

Solution-2: Find another way to get reversed hyps in https://github.com/wenet-e2e/wenet/blob/main/wenet/transformer/asr_model.py#L669-L684 while let ONNX be aware of `beam`

**Comment by xingchensong** [comment]
To verify it, just set `reverseweigt=0.0` in both exporting and decoding. In this way, r_hyps will never be used and hence `Split_XX` will be removed by ONNX.

**Comment by xingchensong** [comment]
I find another thing. `beam` in `decoder_main` means `beam_prune`. It is `nbest` that strictly matches the meaning of `beam` in `export_onnx_cpu`

**Comment by xingchensong** [comment]
> No, I set beam=10, some files can work but this as below: I0515 22:46:56.649355 50924 decoder_main.cc:67] num frames 891 I0515 22:46:56.827426 50924 decoder_main.cc:88] Partial result: 他坚信 I0515 22:46:56.852329 50924 decoder_main.cc:88] Partial result: 他坚信只要 I0515 22:46:56.878832 50924 decoder_main.cc:88] Partial result: 他坚信只要踏实肯 I0515 22:46:56.921988 50924 decoder_main.cc:88] Partial result: 他坚信只要踏实肯干 I0515 22:46:56.951897 50924 decoder_main.cc:88] Partial result: 他坚信只要踏实肯干一定能够 I0515 22:46:56.979951 50924 decoder_main.cc:88] Partial result: 他坚信只要踏实肯干一定能够上好 I0515 22:46:57.022194 50924 decoder_main.cc:88] Partial result: 他坚信只要踏实肯干一定能够上好日子 I0515 22:46:57.051928 50924 decoder_main.cc:81] Rescoring begin 2022-05-15 21:46:57.053215410 [E:onnxruntime:, sequential_executor.cc:368 Execute] Non-zero status code returned while running Split node. Name:'Split_45' Status Message: Cannot split using values in 'split' attribute. Axis=0 Input shape={6} NumOutputs=10 Num entries in 'split' (must equal number of outputs) was 10 Sum of sizes in 'split' (must equal size of selected axis) was 10 terminate called after throwing an instance of 'Ort::Exception' what(): Non-zero status code returned while running Split node. Name:'Split_45' Status Message: Cannot split using values in 'split' attribute. Axis=0 Input shape={6} NumOutputs=10 Num entries in 'split' (must equal number of outputs) was 10 Sum of sizes in 'split' (must equal size of selected axis) was 10

By the way, even we set the correct `nbest` in `decoder_main` and `beam` in `export_onnx_cup`, the error still remains.

**Comment by xingchensong** [comment]
> > I know why.  https://github.com/wenet-e2e/wenet/blob/main/wenet/transformer/asr_model.py#L669-L684 . ONNX treats `for-loop`  statically,  it will always split out 15 tensors if we use `beam=15` when exporting the model.
> > Solution-1: Force padding hyps in https://github.com/wenet-e2e/wenet/blob/main/runtime/core/decoder/onnx_asr_model.cc#L315-L330 if `num_hyps < predefined beamsize`
> > Solution-2: Find another way to get reversed hyps in https://github.com/wenet-e2e/wenet/blob/main/wenet/transformer/asr_model.py#L669-L684 while let ONNX be aware of `beam`
> 
> Prefer to solution2. May We can refine reverse pad sequence using torch.gather

We are open to your PR : )

**Comment by Mddct** [comment]
I'm not very convenient to make a pr at the moment，but  I wrote  some code snippets  that might help
```python

def reverse_hyps(input, seq_len):
  max = torch.max(seq_len)
  index_range = torch.arange(0,max, 1)
  seq_len_expand = seq_len.unsqueeze(1)
  seq_mask = seq_len_expand > index_range
  
  index = (seq_len_expand-1) - index_range
  index = index * (seq_len_expand > index_range)
  
  print(index)
  input = torch.gather(input, 1, index)
  return input * seq_mask


seq_len = torch.tensor([4,5,3])
input = torch.tensor([[1,2,3,4,0], [1,1,1,1,1], [1,2,3,0,0]])
print(reverse_hyps, input)

# tensor([[4, 3, 2, 1, 0],
#         [1, 1, 1, 1, 1],
#         [3, 2, 1, 0, 0]])
```

> > > I know why.  https://github.com/wenet-e2e/wenet/blob/main/wenet/transformer/asr_model.py#L669-L684 . ONNX treats `for-loop`  statically,  it will always split out 15 tensors if we use `beam=15` when exporting the model.
> > > Solution-1: Force padding hyps in https://github.com/wenet-e2e/wenet/blob/main/runtime/core/decoder/onnx_asr_model.cc#L315-L330 if `num_hyps < predefined beamsize`
> > > Solution-2: Find another way to get reversed hyps in https://github.com/wenet-e2e/wenet/blob/main/wenet/transformer/asr_model.py#L669-L684 while let ONNX be aware of `beam`
> > 
> > 
> > Prefer to solution2. May We can refine reverse pad sequence using torch.gather
> 
> We are open to your PR : )

**Comment by Mddct** [comment]
> I'm not very convenient to make a pr at the moment，but I wrote some code snippets that might help
> 
> ```python
> def reverse_pad_list_v2(input, seq_len):
>   max = torch.max(seq_len)
>   index_range = torch.arange(0,max, 1)
>   seq_len_expand = seq_len.unsqueeze(1)
>   seq_mask = seq_len_expand > index_range
>   
>   index = (seq_len_expand-1) - index_range
>   index = index * (seq_len_expand > index_range)
>   
>   input = torch.gather(input, 1, index)
>   return input * seq_mask
> 
> 
> seq_len = torch.tensor([4,5,3])
> input = torch.tensor([[1,2,3,4,0], [1,1,1,1,1], [1,2,3,0,0]])
> print(reverse_pad_list_v2(input, seq_len)
> 
> # tensor([[4, 3, 2, 1, 0],
> #         [1, 1, 1, 1, 1],
> #         [3, 2, 1, 0, 0]])
> ```
> 
> > > > I know why.  https://github.com/wenet-e2e/wenet/blob/main/wenet/transformer/asr_model.py#L669-L684 . ONNX treats `for-loop`  statically,  it will always split out 15 tensors if we use `beam=15` when exporting the model.
> > > > Solution-1: Force padding hyps in https://github.com/wenet-e2e/wenet/blob/main/runtime/core/decoder/onnx_asr_model.cc#L315-L330 if `num_hyps < predefined beamsize`
> > > > Solution-2: Find another way to get reversed hyps in https://github.com/wenet-e2e/wenet/blob/main/wenet/transformer/asr_model.py#L669-L684 while let ONNX be aware of `beam`
> > > 
> > > 
> > > Prefer to solution2. May We can refine reverse pad sequence using torch.gather
> > 
> > 
> > We are open to your PR : )
combine add_sos_eos and reverse_pad_list_v2:
```python
def reverse_pad_list(input, seq_len, eos=10000, ignore_id=0):
  max = torch.max(seq_len)
  index_range = torch.arange(0,max, 1)
  seq_len_expand = seq_len.unsqueeze(1)
  seq_mask = seq_len_expand > index_range
  
  index = (seq_len_expand-1) - index_range
  
  index = index * seq_mask
  print(index)
  input = torch.gather(input, 1, index)
  return torch.where(seq_mask == ignore_id, eos, input)

input = torch.tensor([[10,1,2,3], [10,9,8,4], [10,2,0,0]])
seq_len = torch.tensor([4,4,2])

sos = 10
eos = input[:, 0:1]
input = input[:, 1:]
seq_len = seq_len -1 
r_hyps = torch.cat([eos, reverse_pad_list_v2(input, seq_len)], dim=1)

# tensor([[   10,     3,     2,     1],
#         [   10,     4,     8,     9],
#         [   10,     2, 10000, 10000]])
```
